### ISSUE TITLE
Make SSL/TLS implementation pluggable

### ISSUE THREAD

**Issue by snej** [nim-lang/Nim#14719]
### Summary
Nim's TLS/SSL support is currently hardwired to use OpenSSL. Some programs may want to use alternate APIs (mbedTLS, wolfSSL, bearSSL, SecureTransport, etc.) for reasons of footprint, compatibility, security, etc. This should be possible without having to modify code in the standard library.

### Description
Refactor Nim's socket implementation to split out the OpenSSL-dependent code and put a well-defined API in between, with a mechanism to swap in different implementations of that API, whether in the standard library or application-defined.

### Additional Information
[Forum comment](https://forum.nim-lang.org/t/6442#39741) by leorize:
> the [TLS] implementation is currently integrated with net.Socket. You can open an issue on GitHub requesting a pluggable mechanism and I might look into it. It's not really that hard to replace the current system with a pluggable one, it's just that there aren't enough interest to go for it atm.

**Comment by snej** [comment]
FYI, I'm motivated to work on this. If my team ends up choosing Nim to implement a project I'm speccing, we will probably need to use mbedTLS. I'm new to Nim but have a lot of experience with networking in C/C++.

**Comment by dom96** [comment]
Go for it. Just make sure changes are well tested.

**Comment by alaviss** [comment]
We will probably want #14556 in before getting started on this.

**Comment by FedericoCeratto** [comment]
After OpenSSL, GnuTLS appears to be relatively popular, followed by mbedTLS, according to statistics on Debian.
Yet, supporting multiple implementations, doing extensive testing, and **maintaining them for whole lifetime of the libraries** is quite a big amount of work. Are there other programming languages doing that as part of the stdlib?

**Comment by alaviss** [comment]
We really just need to make it pluggable. External implementations can be provided by 3rd party packages.

**Comment by snej** [comment]
> Are there other programming languages doing that as part of the stdlib?

Go implemented its own TLS package, in keeping with its extreme NIH tendency.

Rust has native_tls which wraps the platform's TLS library, "SChannel on Windows (via the schannel crate), Secure Transport on OSX (via the security-framework crate), and OpenSSL (via the openssl crate) on all other platforms". But it does not appear to be part of the standard library as its repo is outside the Rust org on Github.

**Comment by AZ-X** [comment]
Please support this issue.

**Comment by shirleyquirk** [comment]
per https://github.com/nim-lang/Nim/pull/17741#issuecomment-823928586 there's some tentative support from Araq for something approximating this. To keep momentum, would someone with more skin in the game/experience than I be willing to write an RFC?

**Comment by greengekota** [comment]
I think it would be neat to have libtls API. Applications such as HTTPS can easily (and correctly) be built on top of it.
- An introduction to libtls on youtube: [Rethinking the TLS/SSL API](https://youtube.com/watch?v=Wd_dyRbE4AA)
- libressl has an implementation of those API
- openssl has an implementation called [libretls](https://git.causal.agency/libretls) (confusing name)
- bearssl has an implementation [here](https://github.com/michaelforney/libtls-bearssl)

**Comment by 89z** [comment]
https://github.com/nim-lang/Nim/issues/20313

**Comment by 89z** [comment]
Yes, this is much needed. As was mentioned, Rust has a native TLS crate is available:

https://crates.io/crates/native-tls

which uses OpenSSL on Linux, but SChannel on Windows. Another option would be a pure Nim TLS implementation, if one exists. Here is example again with Rust:

https://crates.io/crates/rustls
